libraries/AP_HAL_Linux/RCOutput_Bebop.cpp

```diff
@@ -0,0 +1,440 @@
+    ret = pthread_cond_init(&_cond, NULL);
+
+    if(ret != 0) {
+        perror("RCout_Bebop: failed to init cond\n");
+        goto err_mutex;
+    }
+
+    ret = pthread_attr_init(&attr);
+    if(ret != 0) {
+        perror("RCOut_Bebop: failed to init attr\n");
+        goto err_mutex;
+    }
+    pthread_attr_setinheritsched(&attr, PTHREAD_EXPLICIT_SCHED);
+    pthread_attr_setschedpolicy(&attr, SCHED_FIFO);
+    pthread_attr_setschedparam(&attr, &param);
+    ret = pthread_create(&_thread, &attr, _control_thread, this);
+    if(ret != 0) {
+        perror("RCOut_Bebop: failed to create thread\n");
+        goto err_mutex;
+    }
+
+    _clear_error();
+
+    /* Set an initial dummy frequency */
+    _frequency = 50;
+    return;
+
+err_mutex:
+    pthread_mutex_unlock(&_mutex);
+    return;
+}
+
+void LinuxRCOutput_Bebop::set_freq(uint32_t chmask, uint16_t freq_hz)
+{
+    _frequency = freq_hz;
+}
+
+uint16_t LinuxRCOutput_Bebop::get_freq(uint8_t ch)
+{
+    return _frequency;
+}
+
+void LinuxRCOutput_Bebop::enable_ch(uint8_t ch)
+{
+}
+
+void LinuxRCOutput_Bebop::disable_ch(uint8_t ch)
+{
+    _stop_prop();
+}
+
+void LinuxRCOutput_Bebop::write(uint8_t ch, uint16_t period_us)
+{
+
+    pthread_mutex_lock(&_mutex);
+    _period_us[ch] = period_us;
+    pthread_cond_signal(&_cond);
+    pthread_mutex_unlock(&_mutex);
+}
+
+void LinuxRCOutput_Bebop::write(uint8_t ch, uint16_t* period_us, uint8_t len)
+{
+    for (int i = 0; i < len; i++)
+        write(ch + i, period_us[i]);
+}
+
+uint16_t LinuxRCOutput_Bebop::read(uint8_t ch)
+{
+    if(ch < BEBOP_BLDC_MOTORS_NUM) {
+        return _period_us[ch];
+    }
+    else {
+        return 0;
+    }
+}
+
+void LinuxRCOutput_Bebop::read(uint16_t* period_us, uint8_t len)
+{
+    for (int i = 0; i < len; i++)
+        period_us[i] = read(0 + i);
+}
+
+void LinuxRCOutput_Bebop::set_esc_scaling(uint16_t min_pwm, uint16_t max_pwm)
+{
+    _min_pwm = min_pwm;
+    _max_pwm = max_pwm;
+}
+
+/* Separate thread to handle the Bebop motors controller */
+void* LinuxRCOutput_Bebop::_control_thread(void *arg) {
+    LinuxRCOutput_Bebop* rcout = (LinuxRCOutput_Bebop *) arg;
+
+    rcout->_run_rcout();
+    return NULL;
+}
+
+void LinuxRCOutput_Bebop::_run_rcout()
+{
+    uint16_t current_period_us[BEBOP_BLDC_MOTORS_NUM];
+    uint8_t i;
+    int ret;
+    struct timespec ts;
+
+    memset(current_period_us, 0, sizeof(current_period_us));
+
+    while(true) {
+        ret = clock_gettime(CLOCK_REALTIME, &ts);
+        if(ret != 0)
+            printf("failed to get time\n");
+
+        if(ts.tv_nsec > (1000000000 - BEBOP_BLDC_TIMEOUT_NS))
+        {
+            ts.tv_sec += 1;
+            ts.tv_nsec = ts.tv_nsec + BEBOP_BLDC_TIMEOUT_NS - 1000000000;
+        }
+        else {
+            ts.tv_nsec += BEBOP_BLDC_TIMEOUT_NS;
+        }
+
+        ret = 0;
+        while((memcmp(_period_us, current_period_us, sizeof(_period_us)) == 0) && (ret == 0))
+            ret = pthread_cond_timedwait(&_cond, &_mutex, &ts);
```

Comment:
The mutex is taken either by init or at the end of the loop after startup.
Moreover, Tridge pointed out that since the writes to uint16_t are atomics, there is even no need for a mutex.